### PR TITLE
Make success property optional

### PR DESCRIPTION
**Pull request by ductiletoaster**
Looking to make the success property optional as it is redundant when validating a response from a client.

### CODE REVIEW

File: src/Http/ErrorResponseBuilder.php [hunk 1]
```diff
@@ -85,11 +85,11 @@ public function addData(array $data):ErrorResponseBuilder
     /**
      * Set the error code and optionally an error message.
      *
-     * @param  string|null       $errorCode
+     * @param  mixed|null       $errorCode
      * @param  string|array|null $message
      * @return self
      */
-    public function setError(string $errorCode = null, $message = null):ErrorResponseBuilder
+    public function setError($errorCode = null, $message = null):ErrorResponseBuilder
```
Comment (flugger) [inline src/Http/ErrorResponseBuilder.php]: Why the change of argument type here?

File: src/Http/ResponseBuilder.php [hunk 2]
```diff
@@ -54,15 +69,19 @@ public function __construct($responseFactory)
      *
      * @param  int|null $statusCode
      * @param  array    $headers
+     * @param  bool     $successFlag
      * @return \Illuminate\Http\JsonResponse
      */
-    public function respond(int $statusCode = null, array $headers = []):JsonResponse
+    public function respond(int $statusCode = null, array $headers = [], bool $successFlag = true):JsonResponse
```
Comment (flugger) [inline src/Http/ResponseBuilder.php]: I don't feel like the flag belongs in the parameter list here, seeing as you can set it using the setter below.

File: src/Http/ResponseBuilder.php [hunk 3]
```diff
@@ -137,12 +169,27 @@ public function jsonSerialize()
...
+     * @return array
+     */
     protected function includeStatusCode(array $data):array
     {
         if (! $this->includeStatusCode) {
             return $data;
         }
 
-        return array_merge(['status' => $this->statusCode], $data);
+        return array_merge([], ['status' => $this->statusCode], $data);
```
Comment (flugger) [inline src/Http/ResponseBuilder.php]: Why are we merging with an empty array here?

File: src/Responder.php [hunk 4]
```diff
@@ -46,20 +46,20 @@ public function __construct(SuccessResponseBuilder $successResponse, ErrorRespon
...
+    public function error($errorCode = null, int $statusCode = null, $message = null):JsonResponse
     {
         if ($exception = config("responder.exceptions.$errorCode")) {
             if (class_exists($exception)) {
                 throw new $exception();
             }
         }
 
-        return $this->errorResponse->setError($errorCode, $message)->respond($statusCode);
+        return $this->errorResponse->setError($errorCode, $message)->respond($statusCode, [], false);
```
Comment (flugger) [inline src/Responder.php]: Why are we overriding the config value for error responses here? I feel like if you enable the success flag you should get it for both success and error responses.

File: src/Http/ErrorResponseBuilder.php (hunk 1 above)
Comment (ductiletoaster) [inline src/Http/ErrorResponseBuilder.php]: It slipped in actually. I was going to move it into a separate PR... but since it is here I had a use case where the errorCode was actually an integer value. 

Example errors.php
`<?php

use App\Enums\ErrorCode;

return [

    /*
    |--------------------------------------------------------------------------
    | Error Message Language Lines
    |--------------------------------------------------------------------------
    |
    | The following language lines are used by the Laravel Responder package.
    | When it generates error responses, it will search the messages array
    | below for any key matching the given error code for the response.
    |
    */
    ErrorCode::UNKNOWN_ERROR => 'Unknown Issue.',

];`

File: src/Http/ResponseBuilder.php (hunk 2 above)
Comment (flugger) [inline src/Http/ResponseBuilder.php]: I guess a better option would be to override the `respond` method in each of the response builders, set the success flag and then call on the parent method?

File: src/Http/ErrorResponseBuilder.php (hunk 1 above)
Comment (flugger) [inline src/Http/ErrorResponseBuilder.php]: Oh, fair enough, we can leave it in, doesn't hurt anyone :)

File: src/Http/ResponseBuilder.php (hunk 3 above)
Comment (ductiletoaster) [inline src/Http/ResponseBuilder.php]: Awe, it is a holder over from other attemps I had made. We can squash this and the other array_merge on `includeSuccessFlag`

File: src/Http/ResponseBuilder.php (hunk 2 above)
Comment (ductiletoaster) [inline src/Http/ResponseBuilder.php]: I actually removed the public setter as it seemed odd to do `setSuccess(true)` on `error()` and vice versa on `success()`

I think it should be a property that is set explicitly within those methods BUT if someone uses response directly they also need to be able to override the default.

Thoughts?

File: src/Http/ErrorResponseBuilder.php (hunk 1 above)
Comment (ductiletoaster) [inline src/Http/ErrorResponseBuilder.php]: Sweet! Haha :)

File: src/Http/ResponseBuilder.php (hunk 2 above)
Comment (flugger) [inline src/Http/ResponseBuilder.php]: Seeing as ResponseBuilder is an abstract class, you must use the success or error builder respectively. I feel like it's a bit strange being able to call `$successResponseBuilder->respond(200, [], false)`, it both is and isn't a success :p 

Instead of overriding the `respond` method, we could also add a protected abstract method `isSuccessResponse` which must return a boolean and be implemented by any child classes.

Comment (ductiletoaster) [inline src/Http/ResponseBuilder.php]: Let me explore that and see what I can put together! Thanks for the feedback

**Comment by flugger** [comment]
The tests still include the third parameter though, might want to remove them as well. I also feel like the `isSuccessResponse` methods should be protected as it doesn't bring any value to the public API in my opinion.

**Comment by flugger** [comment]
I guess the tests are still valid, just without the third parameter and more implicit names, like: `respondMethodShouldSetSuccessFlagToTrue` for SuccessResponseBuilderTest and `respondMethodShouldSetSuccessFlagToFalse` for ErrorResponseBuilderTest. What do you think?

**Comment by ductiletoaster** [comment]
I don't think they are valid any more as isSuccessResponse is a protected method and has not public interface. If anything we should expand the tests to check results based on the config variable being toggeled.

**Comment by flugger** [comment]
Well, we just test the JSON response like in your previous tests, in that sense we test that we get the expected `false` and `true` from each respective response builder implicitly. But yeah, we should probably have an additional test for testing that we dont see a `success` key when the config value is disabled. Something like this perhaps:

```
public function respondMethodShouldNotOutputSuccessFlagWhenDisabled()
{
    $this->app['config']->set('responder.include_success_flag', false);
    $responseBuilder = $this->app->make('responder.success');

    $response = $responseBuilder->respond();

    $this->assertArrayNotHasKey('success', json_decode($response->content(), true));
}
```

**Comment by ductiletoaster** [comment]
I think we could also add more asserts in `respondMethodShouldReturnAJsonResponse` to cover the structure of the json body being returned

**Comment by flugger** [comment]
Yeah indeed, if we add the `respondMethodShouldNotOutputSuccessFlagWhenDisabled` method in both test classes and assert for success => true/false in both `respondMethodShouldReturnAJsonResponse` methods I think we should be set.

**Comment by ductiletoaster** [comment]
Awesome! Just updated

File: tests/Unit/ErrorResponseBuilderTest.php [hunk 5]
```diff
@@ -44,10 +44,29 @@ public function respondMethodShouldReturnAJsonResponse()
...
+
+    /**
+     * Test that the [respond] method does not respond with success flag
+     *
+     * @test
+     */
+    public function respondMethodShouldNotOutputSuccessFlagWhenDisabled()
+    {
+        $this->app['config']->set('responder.include_success_flag', false);
+        $responseBuilder = $this->app->make('responder.success');
```
Comment (flugger) [inline tests/Unit/ErrorResponseBuilderTest.php]: This should be `make('responder.error')` :)

**Comment by flugger** [comment]
Thanks again for all your time put into this, it's highly appreciated! I'll get it merged and tag a release right away.

**Comment by flugger** [comment]
New release tagged: `1.2.20`.

File: tests/Unit/ErrorResponseBuilderTest.php (hunk 5 above)
Comment (ductiletoaster) [inline tests/Unit/ErrorResponseBuilderTest.php]: Nice catch! Might be worth moving all the setup variables into a local `setUp` but I'll update just this little piece for now

Comment (flugger) [inline tests/Unit/ErrorResponseBuilderTest.php]: Already merged and fixed it 🎉

Comment (ductiletoaster) [inline tests/Unit/ErrorResponseBuilderTest.php]: O perfect! Thanks!
